### ISSUE TITLE
Cannot use empty blocks with VRL

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#12778]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

VRL doesn't seem to allow if conditions with empty blocks. One could argue that we maybe _shouldn't_ allow this (though I think we should), but, at the least, the error message could be much better than the one the lexer gives.

### Configuration

```text
[sources.stdin]
type = "stdin"
decoding.codec = "json"

[transforms.remap]
type = "remap"
inputs = [ "stdin" ]
source = '''
.image = string(.image) ?? "unknown"
if starts_with(.image, "foo") {
}
'''

[sinks.my_sink_id]
type = "console"
inputs = [ "remap" ]
encoding.codec = "json"
```

### Version

vector 0.21.1

**Comment by JeanMertz** [comment]
To give some background: if statements are built using (among other things) regular `block` expressions. An empty block expression isn't allowed because the same syntax is used to instantiate an empty object.

That is:

```coffee
foo = {} # empty object
foo = { "foo": true } # object with field `foo`
foo = { true } # block with boolean expression
```

**Comment by JeanMertz** [comment]
Yeah. To be clear, there's nothing preventing us from changing the AST to have a new `EmptyBlock` type, which can then only be used in places where an empty object isn't allowed (such as `if true {}`).

**Comment by spencergilbert** [comment]
> Yeah. To be clear, there's nothing preventing us from changing the AST to have a new `EmptyBlock` type, which can then only be used in places where an empty object isn't allowed (such as `if true {}`).

😆 I do need coffee, as I didn't even consider that - counterpoint, since we do want to limit complexity in VRL (generally) do we even want to support empty blocks?

**Comment by JeanMertz** [comment]
> since we do want to limit complexity in VRL (generally) do we even want to support empty blocks?

That's the reason I never added them, since it serves no purpose, other than trivial examples or "I'm still working on this program, so let me just fill this in later".

I think we should optimize for "this program is ready for production, and there's as little ambiguity and complexity as possible". In the "I'm still working on this" mode, you'll just have to write `if true { null }`, as far as I'm concerned.

I do want to codify this "what we optimize for" better in our design doc (part of the "VRL 2022" doc I'll be tackling soon), to avoid discussions like these, or at least have an agreed-upon starting point for any potential discussion.

**Comment by jszwedko** [comment]
Yeah, we can discuss more as part of the vision doc, but I personally found it surprising that I couldn't have an empty block. This would, as you note, primarily be useful during development. Maybe we can support both modes (perhaps introducing a "strict" mode).
